src/realm/object-store/impl/collection_notifier.cpp

```diff
@@ -51,133 +51,74 @@ CollectionNotifier::get_modification_checker(TransactionChangeInfo const& info,
-    // modified, and if not add it to the stack
-    auto already_checking = [&](int64_t col) {
-        auto end = m_current_path.begin() + depth;
-        auto match = std::find_if(m_current_path.begin(), end, [&](auto& p) {
-            return p.obj_key == obj_key && p.col_key == col;
-        });
-        if (match != end) {
-            for (; match < end; ++match)
-                match->depth_exceeded = true;
+    return any_of(begin(m_callbacks), end(m_callbacks), [](auto callback) {
+        if (callback.key_path_array.size() > 0) {
             return true;
         }
-        m_current_path[depth] = {obj_key, col, false};
-        return false;
-    };
-
-    const Obj obj = table.get_object(ObjKey(obj_key));
-    auto linked_object_changed = [&](OutgoingLink const& link) {
-        if (already_checking(link.col_key))
+        else {
             return false;
-        if (!link.is_list) {
-            if (obj.is_null(ColKey(link.col_key)))
-                return false;
-            auto dst = obj.get<ObjKey>(ColKey(link.col_key)).value;
-            return check_row(*table.get_link_target(ColKey(link.col_key)), dst, depth + 1);
         }
-
-        auto& target = *table.get_link_target(ColKey(link.col_key));
-        auto lvr = obj.get_linklist(ColKey(link.col_key));
-        return std::any_of(lvr.begin(), lvr.end(), [&, this](auto key) {
-            return this->check_row(target, key.value, depth + 1);
-        });
-    };
-
-    return std::any_of(begin(it->links), end(it->links), linked_object_changed);
+    });
 }
 
-bool DeepChangeChecker::check_row(Table const& table, ObjKeyType key, size_t depth)
-{
-    // Arbitrary upper limit on the maximum depth to search
-    if (depth >= m_current_path.size()) {
-        // Don't mark any of the intermediate rows checked along the path as
-        // not modified, as a search starting from them might hit a modification
-        for (size_t i = 0; i < m_current_path.size(); ++i)
-            m_current_path[i].depth_exceeded = true;
-        return false;
-    }
-
-    TableKey table_key = table.get_key();
-    if (depth > 0) {
-        auto it = m_info.tables.find(table_key.value);
-        if (it != m_info.tables.end() && it->second.modifications_contains(key))
-            return true;
-    }
-    auto& not_modified = m_not_modified[table_key.value];
-    auto it = not_modified.find(key);
-    if (it != not_modified.end())
-        return false;
-
-    bool ret = check_outgoing_links(table_key, table, key, depth);
-    if (!ret && (depth == 0 || !m_current_path[depth - 1].depth_exceeded))
-        not_modified.insert(key);
-    return ret;
-}
 
-bool DeepChangeChecker::operator()(ObjKeyType key)
+bool CollectionNotifier::all_callbacks_filtered()
 {
-    if (m_root_object_changes && m_root_object_changes->modifications_contains(key))
-        return true;
-    return check_row(m_root_table, key, 0);
+    return all_of(begin(m_callbacks), end(m_callbacks), [](auto callback) {
```

Comment:
This should take `auto& callback` to avoid copying the callback.

Reply:
Fixed with a0a3569.